packages/ai-chat/src/common/chat-model.ts

```diff
@@ -570,6 +570,16 @@ export class ChangeSetImpl implements ChangeSet {
         return true;
     }
 
+    addOrReplaceElement(element: ChangeSetElement): void {
+        const index = this._elements.findIndex(e => e.uri.toString() === element.uri.toString());
+        if (index >= 0) {
+            this._elements[index] = element;
+        } else {
+            this._elements.push(element);
+        }
+        this.notifyChange();
+    }
```

Comment:
We may want to call `replaceElement()` and only if it returns `false`, call `addElement()` to avoid the code duplication. The only disadvantage would be the duplicate event.

Not tested but would look something like below:

```suggestion
    addOrReplaceElement(element: ChangeSetElement): void {
        if (!this.replaceElement(element)) {
          this.addElement(element);
        }
    }
```

Reply:
@planger Good suggestion, I think we can avoid the duplicate event, see latest commit